test/k8s_job_template_test.go

```diff
@@ -0,0 +1,250 @@
+)
+
+// Test each of the required values. Here, we take advantage of the fact that linter_values.yaml is supposed to define
+// all the required values, so we check the template rendering by nulling out each field.
+func TestK8SJobRequiredValuesAreRequired(t *testing.T) {
+	t.Parallel()
+
+	helmChartPath, err := filepath.Abs(filepath.Join("..", "charts", "k8s-job"))
+	require.NoError(t, err)
+
+	eachRequired := []string{
+		"containerImage.repository",
+		"containerImage.tag",
+		"applicationName",
+	}
+	for _, requiredVal := range eachRequired {
+		// Capture the range value and force it into this scope. Otherwise, it is defined outside this block so it can
+		// change when the subtests parallelize and switch contexts.
+		requiredVal := requiredVal
+		t.Run(requiredVal, func(t *testing.T) {
+			t.Parallel()
+
+			// We make sure to pass in the linter_values.yaml values file, which we assume has all the required values defined.
+			// We then use SetValues to null out the value.
+			options := &helm.Options{
+				ValuesFiles: []string{filepath.Join("..", "charts", "k8s-job", "linter_values.yaml")},
+				SetValues:   map[string]string{requiredVal: "null"},
+			}
+			_, err := helm.RenderTemplateE(t, options, helmChartPath, strings.ToLower(t.Name()), []string{})
+			assert.Error(t, err)
+		})
+	}
+}
+
+// Test each of the optional values defined in linter_values.yaml. Here, we take advantage of the fact that
+// linter_values.yaml is supposed to define all the required values, so we check the template rendering by nulling out
+// each field.
+func TestK8SJobOptionalValuesAreOptional(t *testing.T) {
+	t.Parallel()
+
+	helmChartPath, err := filepath.Abs(filepath.Join("..", "charts", "k8s-job"))
+	require.NoError(t, err)
+
+	eachOptional := []string{
+		"containerImage.pullPolicy",
+	}
+	for _, optionalVal := range eachOptional {
+		// Capture the range value and force it into this scope. Otherwise, it is defined outside this block so it can
+		// change when the subtests parallelize and switch contexts.
+		optionalVal := optionalVal
+		t.Run(optionalVal, func(t *testing.T) {
+			t.Parallel()
+
+			// We make sure to pass in the linter_values.yaml values file, which we assume has all the required values defined.
+			// We then use SetValues to null out the value.
+			options := &helm.Options{
+				ValuesFiles: []string{filepath.Join("..", "charts", "k8s-job", "linter_values.yaml")},
+				SetValues:   map[string]string{optionalVal: "null"},
+			}
+			// Make sure it renders without error
+			helm.RenderTemplate(t, options, helmChartPath, "all", []string{})
+		})
+	}
+}
+
+// Test that annotations render correctly to annotate the Job resource
+func TestK8SJobAnnotationsRenderCorrectly(t *testing.T) {
+	t.Parallel()
+
+	uniqueID := random.UniqueId()
+	// ERROR: Need to find function that can inject annotations into a job
+	job := renderK8SServiceDeploymentWithSetValues(t, map[string]string{"jobAnnotations.unique-id": uniqueID})
```

Comment:
I listened to your outro call, and I think you got a little tripped up about this function here. This helper function is actually defined [right in this directory](https://github.com/Gageperrin/helm-kubernetes-services/blob/feat/k8s-job/test/k8s_service_template_render_helpers_for_test.go#L23) in a helper file. You could follow a similar pattern and also define your helper functions for k8s-job that way. Or if you feel another approach is easier to reason about, that could be great.

Reply:
I can totally understand ramping up on this test as well as Terratest can be a lot in one `go`. (Punintended)